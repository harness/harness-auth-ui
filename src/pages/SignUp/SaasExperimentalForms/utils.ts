/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import CDLogo from "static/images/cd-logo.svg";
import CILogo from "static/images/ci-logo.svg";
import CCMLogo from "static/images/ccm-logo.svg";
import FFLogo from "static/images/ff-logo.svg";
import ChaosLogo from "static/images/chaos-logo.svg";
import HarnessLogo from "static/images/logo.svg";

import PathCD from "static/images/path-cd.svg";
import PathCI from "static/images/path-ci.svg";
import PathCCM from "static/images/path-ccm.svg";
import PathFF from "static/images/path-ff.svg";
import PathChaos from "static/images/path-chaos.svg";
import PathHarness from "static/images/path-harness.svg";

interface ModuleDetailsMap {
  [key: string]: ModuleDetail;
}

interface ModuleDetail {
  logo: string;
  pathImg: string;
  title: string;
  tagLine: string;
  valueProp: string;
  callout: string;
  intro: string;
  cta: string;
  color: string;
}
const MODULE_TO_DETAILS_MAP: ModuleDetailsMap = {
  cd: {
    logo: CDLogo,
    pathImg: PathCD,
    title: "Continuos Delivery",
    tagLine: "Blazing Fast Deployment Pipelines in Minutes",
    valueProp:
      "Self-Service Continuous Delivery that enables engineers to deploy on-demand, without a single script",
    callout:
      "Truatsed by 1000s of Developers, driving 17 Million deployments in last 12 months",
    intro:
      "Define Pipelines that enable you to improve your deployment times, manage all in one place",
    cta: "Let's Deploy",
    color: "#5FB34E"
  },
  chaos: {
    logo: ChaosLogo,
    pathImg: PathChaos,
    title: "Chaos Engineering",
    tagLine: "Blazing Fast Deployment Pipelines in Minutes",
    valueProp:
      "Self-Service Continuous Delivery that enables engineers to deploy on-demand, without a single script",
    callout:
      "Trusted by 1000s of Developers, driving 17 Million deployments in last 12 months",
    intro:
      "Define Pipelines that enable you to improve your deployment times, manage all in one place",
    cta: "Let's Deploy",
    color: "#FF006A"
  },
  ci: {
    logo: CILogo,
    pathImg: PathCI,
    title: "Continuos Integration",
    tagLine: "Harness CI is an enterprise ready, cloud native CI product.",
    valueProp:
      "An intelligent, container-native CI solution with isolated builds and standardized extensions. Build artifacts smarter and faster.",
    callout:
      "Trusted by the New York Times, Roblox, John Deere, eBay and many others to run their build and test suits balzing fast",
    intro:
      "Create a container-based build pipeline in minutes. Even testing is lightweight and fast with Test Intelligence!",
    cta: "Let's build",
    color: "#2BB1F2"
  },
  ccm: {
    logo: CCMLogo,
    pathImg: PathCCM,
    title: "Cloud Cost Management",
    tagLine: "Gain complete cloud control and automated cost optimization ",
    valueProp:
      "Achieve cost transparency and reclaim wasted spend across clouds - with built-in anomaly detection and automation",
    callout:
      "Trusted by developers and finance teams from leading startups to enterprise",
    intro:
      "Manage cloud costs and automate savings from kubernetes pods to databases",
    cta: "Connect Cloud Account",
    color: "#01C9CC"
  },
  ff: {
    logo: FFLogo,
    pathImg: PathFF,
    title: "Feature Flags",
    tagLine: "Confidently deploy features with both velocity and control.",
    valueProp:
      "Ship more features and reduce deployment risk with the control and governance enterprises require - all without breaking things, even while in production.",
    callout: "Trusted by Developers and Product teams from zero to scale.",
    intro:
      "Automate feature releases, build with flags as code, and ship more features with less risk.",
    cta: "Start Managing Features",
    color: "#EE8625"
  },
  default: {
    logo: HarnessLogo,
    pathImg: PathHarness,
    title: "Harness",
    tagLine: "Loved by Developers, Trusted by Businesses",
    valueProp:
      "Harness is the industry's first Software Delivery Platform to use AI to simplify your DevOps processes - CI, CD, Feature Flags, Cloud Costs, and much more.",
    callout:
      "Trusted by Developers, DevOps, DevSecOps, Finance and others, across world's leading digital businesses.",
    intro:
      "Define Pipelines that enable you to improve your deployment times, manage all in one place",
    cta: "Let's Deploy",
    color: "#0278D5"
  }
};

const getModuleDetails = (moduleName: string): ModuleDetail => {
  const details = MODULE_TO_DETAILS_MAP[moduleName];
  if (details) {
    return details;
  }
  return MODULE_TO_DETAILS_MAP.default;
};

export { getModuleDetails };
