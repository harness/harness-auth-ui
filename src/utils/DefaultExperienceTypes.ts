export enum DefaultExperience {
  NG = "NG",
  CG = "CG"
}
