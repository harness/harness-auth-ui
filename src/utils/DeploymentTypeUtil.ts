export const isCommunityPlan = (): boolean =>
  window.deploymentType === "COMMUNITY";

export const isOnPrem = (): boolean => window.deploymentType === "ON_PREM";

export const isSaas = (): boolean => window.deploymentType === "SAAS";
