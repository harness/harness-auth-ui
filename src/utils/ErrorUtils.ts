import toast from "react-hot-toast";

export function handleError(error: any): void {
  const response = error.data;

  if (
    response?.responseMessages?.length > 0 &&
    response?.responseMessages[response.responseMessages.length - 1].message
  ) {
    toast(
      response.responseMessages[response.responseMessages.length - 1].message ||
        ""
    );
  } else {
    toast(error.message);
  }
}
