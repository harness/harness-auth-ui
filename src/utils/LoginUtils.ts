import type { User } from "services/portal";
import AppStorage from "utils/AppStorage";

interface AuthHeader {
  authorization: string;
}

export function formatJWTHeader(authCode: string): AuthHeader {
  const token = window.btoa(
    `${AppStorage.get("twoFactorJwtToken")}:${authCode}`
  );
  const header = {
    authorization: `JWT ${token}`
  };
  return header;
}

export function handleLoginSuccess(resource?: User): void {
  const queryString = window.location.hash?.split("?")?.[1];
  const urlParams = new URLSearchParams(queryString);
  const returnUrl = urlParams?.get("returnUrl");

  if (resource) {
    AppStorage.set("token", resource.token);
    AppStorage.set("uuid", resource.uuid);
    AppStorage.set("acctId", resource.defaultAccountId);
    AppStorage.set("lastTokenSetTime", +new Date());

    if (returnUrl) {
      try {
        const returnUrlObject = new URL(returnUrl);
        // only redirect to same hostname
        if (returnUrlObject.hostname === window.location.hostname) {
          window.location.href = returnUrl;
          return;
        } else {
          // eslint-disable-next-line no-console
          console.warn(
            `"${returnUrl}" is a not a valid redirect due to hostname mismatch`
          );
        }
      } catch (err) {
        // returnUrl is not a valid url. do nothing.
        // eslint-disable-next-line no-console
        console.warn(`"${returnUrl}" is not a valid URL`);
      }
    }

    const experience = resource.accounts?.find(
      (account) => account.uuid === resource.defaultAccountId
    )?.defaultExperience;

    switch (experience) {
      case "NG":
        window.location.href = `/ng/#/account/${resource.defaultAccountId}/projects`;
        return;
      case "CG":
      default:
        window.location.href = `/#/account/${resource.defaultAccountId}/dashboard`;
        return;
    }
  } else {
    // TODO: handle empty login response
  }
}
