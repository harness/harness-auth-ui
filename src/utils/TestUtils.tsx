import React from "react";
import qs from "qs";
import { compile } from "path-to-regexp";
import { Router, Route, Switch } from "react-router-dom";
import { createMemoryHistory } from "history";
import { RestfulProvider, UseGetProps } from "restful-react";
import { queryByAttribute } from "@testing-library/react";

export interface TestWrapperProps {
  path?: string;
  pathParams?: Record<string, string | number>;
  queryParams?: Record<string, unknown>;
}

export const TestWrapper: React.FC<TestWrapperProps> = (props) => {
  const { path = "/", pathParams = {}, queryParams = {} } = props;

  const search = qs.stringify(queryParams, { addQueryPrefix: true });
  const routePath = compile(path)(pathParams) + search;

  const history = React.useMemo(
    () => createMemoryHistory({ initialEntries: [routePath] }),
    []
  );

  return (
    <Router history={history}>
      <RestfulProvider base="/">
        <Switch>
          <Route exact path={path}>
            {props.children}
          </Route>
        </Switch>
      </RestfulProvider>
    </Router>
  );
};

export const queryByNameAttribute = (
  name: string,
  container: HTMLElement
): HTMLElement | null => queryByAttribute("name", container, name);

export type UseGetMockData<
  TData,
  TError = undefined,
  TQueryParams = undefined,
  TPathParams = undefined
> = Required<UseGetProps<TData, TError, TQueryParams, TPathParams>>["mock"];
