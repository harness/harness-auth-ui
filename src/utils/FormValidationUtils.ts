import { regexEmail } from "utils/StringUtils";

export function validateName(name: string): string | undefined {
  if (!name) {
    return "Name is required";
  }
}

export function validateEmail(email: string): string | undefined {
  if (!email) {
    return "An email is required";
  }

  if (!regexEmail.test(email)) {
    return "This email is not valid";
  }
}

export function validatePassword(password: string): string | undefined {
  if (!password) {
    return "A password is required";
  }

  if (password.length < 8 || password.length > 128) {
    return "The password must be between 8 and 128 characters long";
  }
}

export function validatePasswordRequiredOnly(
  password: string
): string | undefined {
  if (!password) {
    return "A password is required";
  }
}
