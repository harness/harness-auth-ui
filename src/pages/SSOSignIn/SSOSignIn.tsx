import React from "react";
import cx from "classnames";
import toast from "react-hot-toast";
import { Form } from "react-final-form";
import { Link } from "react-router-dom";

import { useGetLoginType } from "services/portal";
import RouteDefinitions from "RouteDefinitions";
import BasicLayout from "components/BasicLayout/BasicLayout";
import Field from "components/Field/Field";
import Text from "components/Text/Text";

import logo from "static/images/harness-logo.svg";
import css from "../SignIn/SignIn.module.css";
import { validateEmail } from "utils/FormValidationUtils";

interface SSOFormData {
  email: string;
}

const SSOSignIn: React.FC = () => {
  const { mutate: getLoginType, loading } = useGetLoginType({});

  const handleSSOInit = async (data: SSOFormData): Promise<void> => {
    try {
      const res = await getLoginType({ userName: data.email });
      if (res.resource?.authenticationMechanism === "SAML") {
        if (res.resource.ssorequest?.idpRedirectUrl) {
          window.location.href = res.resource.ssorequest?.idpRedirectUrl;
        }
      } else {
        toast("Single Sign-on is not enabled for your account");
      }
    } catch (err) {
      toast("There was an error");
    }
  };

  return (
    <BasicLayout>
      <div className={cx(css.signin)}>
        <div className={css.header}>
          <img src={logo} width={120} className={css.logo} alt={"Harness"} />
          <div style={{ flex: 1 }} />
          <Link to={RouteDefinitions.toSignIn()}>
            <Text icon="leftArrow">Main Sign In</Text>
          </Link>
        </div>
        <div className={css.title}>
          Sign In with an <br />
          SSO Identity Provider
        </div>
        <div className={css.subtitle}>and get ship done.</div>
        <Form
          onSubmit={handleSSOInit}
          render={({ handleSubmit }) => {
            return (
              <form
                className="layout-vertical spacing-medium"
                onSubmit={handleSubmit}
              >
                <Field
                  name="email"
                  type="email"
                  label="Email"
                  placeholder="email@work.com"
                  disabled={loading}
                  validate={validateEmail}
                />
                <input
                  type="submit"
                  value="Sign In"
                  className="button primary"
                  disabled={loading}
                />
              </form>
            );
          }}
        />
      </div>
    </BasicLayout>
  );
};

export default SSOSignIn;
