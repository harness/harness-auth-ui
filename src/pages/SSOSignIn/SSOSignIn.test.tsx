/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Free Trial 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/05/PolyForm-Free-Trial-1.0.0.txt.
 */

import React from "react";
import { render } from "@testing-library/react";

import SSOSignIn from "./SSOSignIn";
import { TestWrapper } from "utils/TestUtils";

describe("SSOSignIn", () => {
  test("render", () => {
    const { container } = render(
      <TestWrapper path={"/sso-signin"}>
        <SSOSignIn />
      </TestWrapper>
    );
    expect(container).toMatchSnapshot();
  });
});
