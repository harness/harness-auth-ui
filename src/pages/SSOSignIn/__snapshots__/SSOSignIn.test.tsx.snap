// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SSOSignIn render 1`] = `
<div>
  <div
    class="layout"
  >
    <div
      class="cardColumn"
    >
      <div
        class="card"
      >
        <div
          class="signin"
        >
          <div
            class="header"
          >
            <img
              alt="Harness"
              class="logo"
              src="test-file-stub"
              width="120"
            />
            <div
              style="flex: 1;"
            />
            <a
              href="/signin"
            >
              <span
                class="text"
              >
                <img
                  class="icon"
                  height="16"
                  src="test-file-stub"
                  width="16"
                />
                Main sign in
              </span>
            </a>
          </div>
          <div
            class="title"
          >
            Sign in with an 
            <br />
            SSO Identity Provider
          </div>
          <div
            class="subtitle"
          >
            and get ship done.
          </div>
          <form
            class="layout-vertical spacing-medium"
          >
            <div
              class="layout-vertical spacing-small field"
            >
              <label>
                Email
              </label>
              <input
                class=""
                id="email"
                name="email"
                placeholder="email@work.com"
                type="email"
                value=""
              />
            </div>
            <input
              class="button primary"
              type="submit"
              value="Sign in"
            />
          </form>
        </div>
      </div>
    </div>
    <div
      class="imageColumn"
    >
      <img
        alt=""
        aria-hidden="true"
        class="image"
        src="test-file-stub"
      />
    </div>
  </div>
</div>
`;
